sdk/rust/src/http.rs

```diff
@@ -15,13 +17,165 @@ pub use super::wit::wasi::http::types::*;
 /// is no need for streaming bodies.
 pub struct Request {
     /// The method of the request
-    pub method: Method,
-    /// The path together with the query string
-    pub path_and_query: String,
+    method: Method,
+    /// The uri for the request
+    ///
+    /// The first item is set to `None` if the supplied uri is malformed
+    uri: (Option<hyperium::Uri>, String),
     /// The request headers
-    pub headers: Vec<(String, Vec<u8>)>,
+    headers: HashMap<String, HeaderValue>,
     /// The request body as bytes
-    pub body: Vec<u8>,
+    body: Vec<u8>,
+}
+
+enum HeaderValue {
+    String(String),
+    Bytes(Vec<u8>),
+}
+
+impl HeaderValue {
+    /// Turn the `HeaderValue` into bytes
+    fn into_bytes(self) -> Vec<u8> {
+        match self {
+            HeaderValue::String(s) => s.into_bytes(),
+            HeaderValue::Bytes(b) => b,
+        }
+    }
+}
+
+impl Request {
+    fn new(method: Method, uri: impl Into<String>) -> Self {
+        Self {
+            method,
+            uri: Self::parse_uri(uri.into()),
+            headers: HashMap::new(),
+            body: Vec::new(),
+        }
+    }
+
+    /// The request method
+    pub fn method(&self) -> &Method {
+        &self.method
+    }
+
+    /// The request uri
+    pub fn uri(&self) -> &str {
+        &self.uri.1
+    }
+
+    /// The request uri path
+    pub fn path(&self) -> &str {
+        self.uri.0.as_ref().map(|u| u.path()).unwrap_or_default()
+    }
+
+    /// The request uri query
+    pub fn query(&self) -> &str {
+        self.uri
+            .0
+            .as_ref()
+            .and_then(|u| u.query())
+            .unwrap_or_default()
+    }
+
+    /// The request headers
+    ///
+    /// This only returns headers that are utf8 encoded
+    pub fn headers(&self) -> impl Iterator<Item = (&str, &str)> {
+        self.headers.iter().filter_map(|(k, v)| match v {
+            HeaderValue::String(v) => Some((k.as_str(), v.as_str())),
+            HeaderValue::Bytes(_) => None,
+        })
+    }
+
+    /// Return a header value
+    ///
+    /// Will return `None` if the header does not exist or if it is not utf8
+    pub fn header(&self, name: &str) -> Option<&str> {
```

Comment:
This should be case-insensitive, either by downcasing everywhere or by explicitly searching.